src/fsharp/TypeChecker.fs

```diff
@@ -8297,30 +8297,84 @@ and delayRest rest mPrior delayed =
 //------------------------------------------------------------------------- 
 
 and TcFunctionApplicationThen cenv overallTy env tpenv mExprAndArg expr exprty (synArg: SynExpr) atomicFlag delayed = 
-    
     let denv = env.DisplayEnv
     let mArg = synArg.Range
     let mFunExpr = expr.Range
+
+    /// Finds last ident of LongIdent in SynExpr.Ident, LongIdent or TypeApp.
+    /// Type checkes the whole thing as it goes in order to:
+    ///
+    /// * ensure we pass well typed things to `nameof`
+    ///
+    /// * not to loose `FSharpSymbolUse` for `nameof` argument, because we erase it with `Expr.Const(Const.String ...)` further in this function.
+    let (|LastPartOfLongIdentStripParens|_|) expr =
+        let rec findIdents expr =
+            match expr with
+            | SynExpr.Ident ident -> Some ident
+            | SynExpr.TypeApp (expr = expr) -> findIdents expr
+            | SynExpr.LongIdent(_, LongIdentWithDots(idents, _), _, _) -> List.tryLast idents
+            | _ -> None
+        findIdents expr
+
+    let tcExpr = function
```

Comment:
Should this function be called ``TcNameOfExpr``?  I think it's only called from the ``nameof`` branch of the code, correct?

If so, I think it's best to factor it out to be a module-level function definition called ``TcNameOfExpr``